opentracing/propagation.py

```diff
@@ -0,0 +1,169 @@
+
+    See Extractor and Tracer.injector().
+    """
+
+    def inject_span(self, span, carrier):
+        """inject_span takes `span` and injects it into `carrier`.
+
+        The actual type of `carrier` depends on the `format` value passed to
+        `Tracer.injector()`.
+
+        Implementations may raise opentracing.InvalidCarrierException or any
+        other implementation-specific exception if injection fails.
+
+        :param span: the Span instance to inject
+        :param carrier: the format-specific carrier object to inject into
+        """
+        pass
+
+
+class Extractor(object):
+    """An Extractor extracts Span instances from a format-specific "carrier"
+    object.
+
+    Typically the carrier has just arrived across a process boundary, often via
+    an RPC (though message queues and other IPC mechanisms are also reasonable
+    places to use an Extractor).
+
+    See Injector and Tracer.extractor().
+    """
+
+    def join_trace(self, operation_name, carrier):
+        """join_trace returns a Span instance with operation name `operation_name`
+        that's joined to the trace state embedded within `carrier`, or None if
+        no such trace state could be found.
+
+        Implementations may raise opentracing.InvalidCarrierException,
+        opentracing.TraceCorruptedException, or implementation-specific errors
+        if there are more fundamental problems with `carrier`.
+
+        Upon success, the returned Span instance is already started.
+
+        :param operation_name: the operation name for the returned Span (which
+            can be set later via Span.set_operation_name())
+        :param carrier: the format-specific carrier object to extract from
+
+        :return: a Span instance joined to the trace state in `carrier` or None
+            if no such trace state could be found.
+        """
+        return None
+
+
+class InvalidCarrierException(Exception):
+    """InvalidCarrierException should be used when the provided carrier
+    instance does not match what the `format` argument requires.
+
+    See Injector and Extractor.
+    """
+    pass
+
+
+class TraceCorruptedException(Exception):
+    """TraceCorruptedException should be used when the underlynig trace state
+    is seemingly present but not well-formed.
+
+    See Extractor.join_trace.
+    """
+    pass
+
+
+class Format:
+    """A namespace for builtin Injector/Extractor formats.
+
+    These static constants are intended for use in the Tracer.injector() and
+    Tracer.extractor() methods. E.g.,
+
+        tracer.injector(Format.SPLIT_BINARY).inject_span(...)
+
+    """
+
+    # The SPLIT_BINARY format pairs with an Injector or Extractor that expects
+    # a SplitBinaryCarrier carrier object.
+    SPLIT_BINARY = "split_binary"
+
+    # The SPLIT_TEXT format pairs with an Injector or Extractor that expects a
+    # SplitTextCarrier carrier object.
+    SPLIT_TEXT = "split_text"
+
+
+class SplitBinaryCarrier(object):
```

Comment:
I had to add these carrier types (parallel to Go)... I initially did the namedtuple thing, but they do funky things with `__new__` and I wasn't sure it was worth the hackery I was going to need in order to support `SplitBinaryCarrier()` as a valid (no-arg) way to construct an empty instance.